rclcpp/include/rclcpp/function_traits.hpp

```diff
@@ -96,6 +96,24 @@ struct function_traits<
   : function_traits<ReturnTypeT(Args ...)>
 {};
 
+// std::bind for object const methods
+template<typename ClassT, typename ReturnTypeT, typename ... Args, typename ... FArgs>
+#if defined _LIBCPP_VERSION  // libc++ (Clang)
+struct function_traits<std::__bind<ReturnTypeT (ClassT::*)(Args ...), FArgs ...>>
+#elif defined _GLIBCXX_RELEASE  // glibc++ (GNU C++ >= 7.1)
+struct function_traits<std::_Bind<ReturnTypeT(ClassT::*(FArgs ...))(Args ...) const>>
```

Comment:
@esteve Nothing questioning any of this, but could you explain to me what exactly this is? I can tell it's the result of a `std::bind` (which makes this compiler dependent), but I'm not following the two variadic templates here at all. If you have time to explain, I'd love to learn why this patch is the fix.

Reply:
Sure, no problem. Let's say we have this struct:

```c++
struct A {
    int member_const(float x, double y) const {
        (void)x;
        (void)y;
        return 1;
    }

    int member_non_const(float x, double y) {
        (void)x;
        (void)y;
        return 2;
    }
};
```

They are the same, but one of them is declared as `const`. So because of the slightly different signature, the template specializations we have for `std::bind` don't match. Its real signature for glibc++ and GCC >= 7.1 is the following:

```c++
std::_Bind<int (A::*(A*, std::_Placeholder<1>, std::_Placeholder<2>))(float, double) const>
```

For which we don't have any match in `function_traits`, the fix consists in duplicating the signatures we have for object members and just add a `const` where needed. However, I don't have a macOS or Windows machine at hand, so I haven't updated the signatures for them. To fix it, we just need to run the `function_traits` testsuite, and update the specializations to match the signatures.

Reply:
@esteve Ah, that makes a lot of sense. Thank you very much!

Reply:
@esteve I ran it on my windows machine for you: https://gist.github.com/allenh1/0b879d6a28c831615c68f307934a5e0f

Not building at the moment, it seems. But this should have all you need in it! Let me know if you need anything else.

Reply:
@allenh1 awesome, thanks!

Reply:
@allenh1 I justed updated the specialization with the info you pasted, could you try it again? Thanks.

Reply:
@esteve I _think_ it's fixed. I have a failure, but my build is not up to date yet (and it is unrelated to function_traits). I'll build my workspace and let you know if it fails.

Reply:
@esteve Definitely working on my machine